src/dune_lang/package_name.ml

```diff
@@ -49,11 +49,24 @@ module Opam_compatible = struct
 
     let of_string_opt s = Option.some_if (is_valid_string s) s
 
-    let make_valid s =
+    let escape s =
       let replaced = String.map s ~f:(fun c -> if is_valid_char c then c else '_') in
       if is_valid_string replaced then replaced else "p" ^ replaced
     ;;
 
+    let make_valid s =
+      match index_opt s '.' with
+      | None -> escape s
+      | Some i ->
+        let before = sub s ~pos:0 ~len:i in
+        let s_len = length s in
+        if is_valid_string before && s_len > i + 1
+        then (
+          let after = sub s ~pos:(i + 1) ~len:(s_len - i - 1) in
+          Printf.sprintf "(%s (= %s))" before after)
```

Comment:
Just a stylistic convention, we don't usually use `Printf.sprintf` in dune, it is already included so `sprintf` should suffice.